File: apps/api-go/internal/blog/blog.go
package blog

import (
	"crypto/rand"
	"encoding/hex"
	"errors"
	"fmt"
	"net/http"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"

	"github.com/xirothedev/webdevstudios/apps/api-go/internal/web"
	"gorm.io/gorm"

	"strings"

	"github.com/xirothedev/webdevstudios/apps/api-go/internal/cart"
	"github.com/xirothedev/webdevstudios/apps/api-go/internal/storage"
)

// Post mirrors Prisma BlogPost. The markdown body itself lives in R2
// (contentUrl); the DB row is metadata only.
type Post struct {
	ID              string     `gorm:"column:id"`
	Slug            string     `gorm:"column:slug"`
	Title           string     `gorm:"column:title"`
	ContentKey      string     `gorm:"column:contentKey"`
	ContentSize     *int       `gorm:"column:contentSize"`
	Excerpt         *string    `gorm:"column:excerpt"`
	CoverImage      *string    `gorm:"column:coverImage"`
	AuthorID        string     `gorm:"column:authorId"`
	IsPublished     bool       `gorm:"column:isPublished"`
	PublishedAt     *time.Time `gorm:"column:publishedAt"`
	ViewCount       int        `gorm:"column:viewCount"`
	MetaTitle       *string    `gorm:"column:metaTitle"`
	MetaDescription *string    `gorm:"column:metaDescription"`
	CreatedAt       time.Time  `gorm:"column:createdAt"`
	UpdatedAt       time.Time  `gorm:"column:updatedAt"`

	Author Author `gorm:"foreignKey:AuthorID;references:ID"`
}

func (Post) TableName() string { return "blog_posts" }

type Author struct {
	ID       string  `gorm:"column:id"`
	FullName *string `gorm:"column:fullName"`
	Avatar   *string `gorm:"column:avatar"`
}

func (Author) TableName() string { return "users" }

type PostDTO struct {
	ID              string     `json:"id"`
	Slug            string     `json:"slug"`
	Title           string     `json:"title"`
	ContentKey      string     `json:"contentKey"`
	Content         string     `json:"content,omitempty"`
	ContentSize     *int       `json:"contentSize"`
	Excerpt         *string    `json:"excerpt"`
	CoverImage      *string    `json:"coverImage"`
	Author          AuthorDTO  `json:"author"`
	IsPublished     bool       `json:"isPublished"`
	PublishedAt     *time.Time `json:"publishedAt"`
	ViewCount       int        `json:"viewCount"`
	MetaTitle       *string    `json:"metaTitle"`
	MetaDescription *string    `json:"metaDescription"`
	CreatedAt       time.Time  `json:"createdAt"`
	UpdatedAt       time.Time  `json:"updatedAt"`
}

type AuthorDTO struct {
	ID       string  `json:"id"`
	FullName *string `json:"fullName"`
	Avatar   *string `json:"avatar"`
}

type ListDTO struct {
	Posts []PostDTO `json:"posts"`
	Total int64     `json:"total"`
}

func toDTO(p Post) PostDTO {
	return PostDTO{
		ID: p.ID, Slug: p.Slug, Title: p.Title,
		ContentKey: p.ContentKey, ContentSize: p.ContentSize,
		Excerpt: p.Excerpt, CoverImage: p.CoverImage,
		Author:      AuthorDTO{ID: p.Author.ID, FullName: p.Author.FullName, Avatar: p.Author.Avatar},
		IsPublished: p.IsPublished, PublishedAt: p.PublishedAt, ViewCount: p.ViewCount,
		MetaTitle: p.MetaTitle, MetaDescription: p.MetaDescription,
		CreatedAt: p.CreatedAt, UpdatedAt: p.UpdatedAt,
	}
}

type Handler struct {
	db    *gorm.DB
	store *storage.Service
}

func Register(v1 *gin.RouterGroup, db *gorm.DB, authRequired, adminRequired gin.HandlerFunc, store *storage.Service) {
	h := &Handler{db, store}
	g := v1.Group("/blog/posts")
	g.GET("", h.list)          // published only
	g.GET("/search", h.search) // published only
	g.GET("/admin/all", authRequired, adminRequired, h.listAll)
	g.POST("", authRequired, adminRequired, h.create)
	g.PATCH("/:id", authRequired, adminRequired, h.update)
	g.DELETE("/:id", authRequired, adminRequired, h.remove)
	g.GET("/:slug", h.getBySlug)
}

func (h *Handler) list(c *gin.Context) {
	page, size := paging(c)
	where := map[string]any{"isPublished": true}
	var total int64
	if err := h.db.Model(&Post{}).Where(where).Count(&total).Error; err != nil {
		fail(c, err)
		return
	}
	var rows []Post
	if err := h.db.Preload("Author").Where(where).
		Order(`"publishedAt" DESC`).Offset((page - 1) * size).Limit(size).Find(&rows).Error; err != nil {
		fail(c, err)
		return
	}
	c.JSON(http.StatusOK, toList(rows, total))
}

func (h *Handler) listAll(c *gin.Context) {
	page, size := paging(c)
	var total int64
	if err := h.db.Model(&Post{}).Count(&total).Error; err != nil {
		fail(c, err)
		return
	}
	var rows []Post
	if err := h.db.Preload("Author").
		Order(`"publishedAt" DESC`).Offset((page - 1) * size).Limit(size).Find(&rows).Error; err != nil {
		fail(c, err)
		return
	}
	c.JSON(http.StatusOK, toList(rows, total))
}

func (h *Handler) search(c *gin.Context) {
	q := c.Query("q")
	if q == "" {
		nest(c, http.StatusBadRequest, "q is required")
		return
	}
	page, size := paging(c)
	pattern := "%" + q + "%"
	base := h.db.Model(&Post{}).Where(map[string]any{"isPublished": true}).
		Where(`title ILIKE ? OR excerpt ILIKE ?`, pattern, pattern)
	var total int64
	if err := base.Count(&total).Error; err != nil {
		fail(c, err)
		return
	}
	var rows []Post
	if err := h.db.Preload("Author").Where(map[string]any{"isPublished": true}).
		Where(`title ILIKE ? OR excerpt ILIKE ?`, pattern, pattern).
		Order(`"publishedAt" DESC`).Offset((page - 1) * size).Limit(size).Find(&rows).Error; err != nil {
		fail(c, err)
		return
	}
	c.JSON(http.StatusOK, toList(rows, total))
}

func (h *Handler) getBySlug(c *gin.Context) {
	slug := c.Param("slug")
	var p Post
	if err := h.db.Preload("Author").First(&p, map[string]any{"slug": slug}).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			nest(c, http.StatusNotFound, fmt.Sprintf("Blog post with slug %q not found", slug))
			return
		}
		fail(c, err)
		return
	}
	if c.Query("includeContent") == "true" {
		body, err := h.store.GetObject(p.ContentKey)
		if err != nil {
			if errors.Is(err, storage.ErrUnavailable) {
				nest(c, http.StatusNotImplemented, "includeContent requires R2_* env vars")
				return
			}
			nest(c, http.StatusInternalServerError, err.Error())
			return
		}
		dto := toDTO(p)
		dto.Content = string(body)
		c.JSON(http.StatusOK, dto)
		return
	}
	includeContent := false
	if includeContent {
		// Content bytes live in R2; fetching lands with the storage phase.
		nest(c, http.StatusNotImplemented, "includeContent requires storage integration (Phase 7)")
		return
	}
	if p.IsPublished {
		h.db.Model(&Post{}).Where(map[string]any{"id": p.ID}).
			UpdateColumn("viewCount", gorm.Expr(`"viewCount" + ?`, 1))
	}
	c.JSON(http.StatusOK, toDTO(p))
}

func toList(rows []Post, total int64) ListDTO {
	out := ListDTO{Posts: make([]PostDTO, 0, len(rows)), Total: total}
	for _, r := range rows {
		out.Posts = append(out.Posts, toDTO(r))
	}
	return out
}

func paging(c *gin.Context) (int, int) {
	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	size, _ := strconv.Atoi(c.DefaultQuery("pageSize", "10"))
	if page < 1 {
		page = 1
	}
	if size < 1 || size > 100 {
		size = 10
	}
	return page, size
}

func fail(c *gin.Context, err error) {
	var he *cart.HTTPError
	if errors.As(err, &he) {
		nest(c, he.Code, he.Msg)
		return
	}
	nest(c, http.StatusInternalServerError, err.Error())
}

func nest(c *gin.Context, code int, msg string) {
	c.JSON(code, gin.H{"statusCode": code, "message": msg, "error": http.StatusText(code)})
}

func (h *Handler) create(c *gin.Context) {
	var in struct {
		Slug            string  `json:"slug" binding:"required"`
		Title           string  `json:"title" binding:"required,max=255"`
		Content         string  `json:"content" binding:"required"`
		Excerpt         *string `json:"excerpt"`
		CoverImage      *string `json:"coverImage"`
		IsPublished     *bool   `json:"isPublished"`
		MetaTitle       *string `json:"metaTitle"`
		MetaDescription *string `json:"metaDescription"`
	}
	if !web.Bind(c, &in) {
		return
	}
	var count int64
	h.db.Model(&Post{}).Where(map[string]any{"slug": in.Slug}).Count(&count)
	if count > 0 {
		nest(c, http.StatusConflict, fmt.Sprintf("Blog post with slug %q already exists", in.Slug))
		return
	}
	excerpt := autoExcerpt(in.Content)
	if in.Excerpt != nil {
		excerpt = *in.Excerpt
	}
	published := false
	if in.IsPublished != nil {
		published = *in.IsPublished
	}
	size := len(in.Content)
	p := Post{
		ID: newID(), Slug: in.Slug, Title: in.Title, ContentKey: "",
		ContentSize: &size, Excerpt: &excerpt, CoverImage: in.CoverImage,
		AuthorID: c.GetString("userId"), IsPublished: published,
		MetaTitle: in.MetaTitle, MetaDescription: in.MetaDescription,
	}
	if published {
		now := time.Now()
		p.PublishedAt = &now
	}
	if err := h.db.Create(&p).Error; err != nil {
		fail(c, err)
		return
	}
	key := "blog/posts/" + p.ID + "/content.md"
	if err := h.store.PutObject(key, []byte(in.Content), "text/markdown"); err != nil {
		h.db.Delete(&Post{}, map[string]any{"id": p.ID}) // mirror Nest rollback
		status := http.StatusInternalServerError
		if errors.Is(err, storage.ErrUnavailable) {
			status = http.StatusNotImplemented
		}
		nest(c, status, err.Error())
		return
	}
	h.db.Model(&Post{}).Where(map[string]any{"id": p.ID}).Update("contentKey", key)
	p.ContentKey = key
	h.db.Preload("Author").First(&p, map[string]any{"id": p.ID})
	c.JSON(http.StatusCreated, toDTO(p))
}

func (h *Handler) update(c *gin.Context) {
	id := c.Param("id")
	var p Post
	if err := h.db.Preload("Author").First(&p, map[string]any{"id": id}).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			nest(c, http.StatusNotFound, "Blog post not found")
			return
		}
		fail(c, err)
		return
	}
	var in struct {
		Title           *string `json:"title"`
		Content         *string `json:"content"`
		Excerpt         *string `json:"excerpt"`
		CoverImage      *string `json:"coverImage"`
		IsPublished     *bool   `json:"isPublished"`
		MetaTitle       *string `json:"metaTitle"`
		MetaDescription *string `json:"metaDescription"`
	}
	if !web.Bind(c, &in) {
		return
	}
	updates := map[string]any{}
	if in.Title != nil {
		updates["title"] = *in.Title
	}
	if in.Excerpt != nil {
		updates["excerpt"] = *in.Excerpt
	}
	if in.CoverImage != nil {
		updates["coverImage"] = *in.CoverImage
	}
	if in.MetaTitle != nil {
		updates["metaTitle"] = *in.MetaTitle
	}
	if in.MetaDescription != nil {
		updates["metaDescription"] = *in.MetaDescription
	}
	if in.IsPublished != nil && *in.IsPublished && p.PublishedAt == nil {
		updates["publishedAt"] = time.Now()
	}
	if in.IsPublished != nil {
		updates["isPublished"] = *in.IsPublished
	}
	if len(updates) > 0 {
		if err := h.db.Model(&Post{}).Where(map[string]any{"id": id}).Updates(updates).Error; err != nil {
			fail(c, err)
			return
		}
	}
	if in.Content != nil {
		if err := h.store.PutObject(p.ContentKey, []byte(*in.Content), "text/markdown"); err != nil {
			status := http.StatusInternalServerError
			if errors.Is(err, storage.ErrUnavailable) {
				status = http.StatusNotImplemented
			}
			nest(c, status, err.Error())
			return
		}
		size := len(*in.Content)
		h.db.Model(&Post{}).Where(map[string]any{"id": id}).Update("contentSize", size)
	}
	h.db.Preload("Author").First(&p, map[string]any{"id": id})
	c.JSON(http.StatusOK, toDTO(p))
}

func (h *Handler) remove(c *gin.Context) {
	id := c.Param("id")
	var p Post
	if err := h.db.First(&p, map[string]any{"id": id}).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			nest(c, http.StatusNotFound, "Blog post not found")
			return
		}
		fail(c, err)
		return
	}
	if err := h.db.Delete(&Post{}, map[string]any{"id": id}).Error; err != nil {
		fail(c, err)
		return
	}
	_ = h.store.DeleteObject(p.ContentKey) // object errors never fail the delete
	c.JSON(http.StatusOK, gin.H{"success": true})
}

// autoExcerpt mirrors extractExcerpt loosely: first ~200 chars, markdown stripped.
func autoExcerpt(content string) string {
	s := strings.NewReplacer("#", "", "*", "", "`", "", "\n", " ").Replace(content)
	for len(s) > 3 && s[0] == ' ' {
		s = s[1:]
	}
	if len(s) > 200 {
		s = s[:200]
	}
	return s
}

func newID() string {
	b := make([]byte, 12)
	_, _ = rand.Read(b)
	return hex.EncodeToString(b)
}

File: apps/api-effect/src/routes/blog.ts
import { Schema } from 'effect';
import { HttpApi, HttpApiBuilder, HttpApiEndpoint, HttpApiGroup } from 'effect/unstable/httpapi';

import { wrap, bodyOf } from '../lib/http';
import type { BlogPost, User } from '../generated/prisma/client';
import { ApiError } from '../lib/errors';
import type { DatabaseClient } from '../lib/prisma';
import { bindBody } from '../lib/validate';
import { requireAdmin } from '../lib/auth';
import { goTime, newId, paging } from '../lib/util';
import { getObject, putObject, deleteObject, StorageUnavailableError } from '../lib/storage';

type PostRow = BlogPost & { author: User };

function toDTO(p: PostRow, content?: string) {
  return {
    id: p.id,
    slug: p.slug,
    title: p.title,
    contentKey: p.contentKey,
    ...(content !== undefined ? { content } : {}),
    contentSize: p.contentSize,
    excerpt: p.excerpt,
    coverImage: p.coverImage,
    author: { id: p.author.id, fullName: p.author.fullName, avatar: p.author.avatar },
    isPublished: p.isPublished,
    publishedAt: goTime(p.publishedAt),
    viewCount: p.viewCount,
    metaTitle: p.metaTitle,
    metaDescription: p.metaDescription,
    createdAt: goTime(p.createdAt),
    updatedAt: goTime(p.updatedAt),
  };
}

// ponytail: mirrors Go extractExcerpt loosely — first ~200 chars, markdown stripped.
function autoExcerpt(content: string): string {
  let s = content.replace(/[#*`]/g, '').replace(/\n/g, ' ');
  while (s.length > 3 && s[0] === ' ') s = s.slice(1);
  if (s.length > 200) s = s.slice(0, 200);
  return s;
}

// Go uses plain `publishedAt DESC` — Postgres nulls-first, so draft posts sort first.
const ORDER = { publishedAt: 'desc' as const };

async function fetchPost(
  db: DatabaseClient,
  idOrSlug: 'id' | 'slug',
  value: string,
): Promise<PostRow> {
  const p = await db.blogPost.findFirst({
    where: idOrSlug === 'id' ? { id: value } : { slug: value },
    include: { author: true },
  });
  if (p === null) {
    throw new ApiError(
      404,
      idOrSlug === 'slug' ? `Blog post with slug "${value}" not found` : 'Blog post not found',
    );
  }
  return p;
}

const IdParams = Schema.Struct({ id: Schema.String });
const SlugParams = Schema.Struct({ slug: Schema.String });

// Field order mirrors toDTO — the success codec renders the 200 response body.
const PostDto = Schema.Struct({
  id: Schema.String,
  slug: Schema.String,
  title: Schema.String,
  contentKey: Schema.String,
  content: Schema.optional(Schema.String),
  contentSize: Schema.NullOr(Schema.Number),
  excerpt: Schema.NullOr(Schema.String),
  coverImage: Schema.NullOr(Schema.String),
  author: Schema.Struct({
    id: Schema.String,
    fullName: Schema.NullOr(Schema.String),
    avatar: Schema.NullOr(Schema.String),
  }),
  isPublished: Schema.Boolean,
  publishedAt: Schema.NullOr(Schema.String),
  viewCount: Schema.Number,
  metaTitle: Schema.NullOr(Schema.String),
  metaDescription: Schema.NullOr(Schema.String),
  createdAt: Schema.NullOr(Schema.String),
  updatedAt: Schema.NullOr(Schema.String),
});

const PostListDto = Schema.Struct({ posts: Schema.Array(PostDto), total: Schema.Number });
const SuccessDto = Schema.Struct({ success: Schema.Boolean });

export const blogGroup = HttpApiGroup.make('blog').add(
  HttpApiEndpoint.get('listPosts', '/v1/blog/posts', { success: PostListDto }),
  HttpApiEndpoint.get('searchPosts', '/v1/blog/posts/search', { success: PostListDto }),
  HttpApiEndpoint.get('listAllPosts', '/v1/blog/posts/admin/all', { success: PostListDto }),
  HttpApiEndpoint.post('createPost', '/v1/blog/posts', { success: PostDto }),
  HttpApiEndpoint.patch('updatePost', '/v1/blog/posts/:id', { success: PostDto, params: IdParams }),
  HttpApiEndpoint.delete('deletePost', '/v1/blog/posts/:id', {
    success: SuccessDto,
    params: IdParams,
  }),
  HttpApiEndpoint.get('getPost', '/v1/blog/posts/:slug', { success: PostDto, params: SlugParams }),
);

export const blogLocal = HttpApi.make('api-effect').add(blogGroup);

export const blogHandlers = HttpApiBuilder.group(blogLocal, 'blog', (h) =>
  h
    .handle(
      'listPosts',
      wrap(true, async (ctx) => {
        // ponytail: Go blog paging reads `pageSize`, not `limit`.
        const { page, limit } = paging(ctx.query.page, ctx.query.pageSize, 100);
        const where = { isPublished: true };
        const total = await ctx.db.blogPost.count({ where });
        const rows = await ctx.db.blogPost.findMany({
          where,
          include: { author: true },
          orderBy: ORDER,
          skip: (page - 1) * limit,
          take: limit,
        });
        return { posts: rows.map((r) => toDTO(r)), total };
      }),
    )
    .handle(
      'searchPosts',
      wrap(true, async (ctx) => {
        const q = ctx.query.q;
        if (!q) throw new ApiError(400, 'q is required');
        const { page, limit } = paging(ctx.query.page, ctx.query.pageSize, 100);
        const where = {
          isPublished: true,
          OR: [
            { title: { mode: 'insensitive' as const, contains: q } },
            { excerpt: { mode: 'insensitive' as const, contains: q } },
          ],
        };
        const total = await ctx.db.blogPost.count({ where });
        const rows = await ctx.db.blogPost.findMany({
          where,
          include: { author: true },
          orderBy: ORDER,
          skip: (page - 1) * limit,
          take: limit,
        });
        return { posts: rows.map((r) => toDTO(r)), total };
      }),
    )
    .handle(
      'listAllPosts',
      wrap(true, async (ctx) => {
        await requireAdmin(ctx);
        const { page, limit } = paging(ctx.query.page, ctx.query.pageSize, 100);
        const total = await ctx.db.blogPost.count({});
        const rows = await ctx.db.blogPost.findMany({
          include: { author: true },
          orderBy: ORDER,
          skip: (page - 1) * limit,
          take: limit,
        });
        return { posts: rows.map((r) => toDTO(r)), total };
      }),
    )
    .handle(
      'createPost',
      wrap(true, async (ctx) => {
        const auth = await requireAdmin(ctx);
        const in1 = bindBody(await bodyOf(ctx), {
          Slug: { type: 'string', required: true },
          Title: { type: 'string', required: true, maxLen: 255 },
          Content: { type: 'string', required: true },
          Excerpt: { type: 'string' },
          CoverImage: { type: 'string' },
          IsPublished: { type: 'boolean' },
          MetaTitle: { type: 'string' },
          MetaDescription: { type: 'string' },
        });
        const slug = in1.slug;
        const dup = await ctx.db.blogPost.count({ where: { slug } });
        if (dup > 0) {
          throw new ApiError(409, `Blog post with slug "${slug}" already exists`);
        }
        let excerpt = autoExcerpt(in1.content);
        if (typeof in1.excerpt === 'string') excerpt = in1.excerpt;
        const published = in1.isPublished === true;
        const id = newId();
        const created = await ctx.db.blogPost.create({
          data: {
            id,
            slug,
            title: in1.title,
            contentKey: '',
            contentSize: in1.content.length,
            excerpt,
            coverImage: in1.coverImage ?? null,
            authorId: auth.user.id,
            isPublished: published,
            publishedAt: published ? new Date() : null,
            metaTitle: in1.metaTitle ?? null,
            metaDescription: in1.metaDescription ?? null,
          },
          include: { author: true },
        });
        const key = `blog/posts/${id}/content.md`;
        try {
          await putObject(key, new TextEncoder().encode(in1.content), 'text/markdown');
        } catch (e) {
          await ctx.db.blogPost.delete({ where: { id } }).catch(() => {});
          throw e instanceof StorageUnavailableError
            ? new ApiError(501, e.message)
            : new ApiError(500, e instanceof Error ? e.message : String(e));
        }
        await ctx.db.blogPost.update({ where: { id }, data: { contentKey: key } });
        ctx.setStatus(201);
        return toDTO({ ...created, contentKey: key });
      }),
    )
    .handle(
      'updatePost',
      wrap(true, async (ctx) => {
        await requireAdmin(ctx);
        const p = await fetchPost(ctx.db, 'id', ctx.param('id'));
        const in1 = bindBody(await bodyOf(ctx), {
          Title: { type: 'string' },
          Content: { type: 'string' },
          Excerpt: { type: 'string' },
          CoverImage: { type: 'string' },
          IsPublished: { type: 'boolean' },
          MetaTitle: { type: 'string' },
          MetaDescription: { type: 'string' },
        });
        // Go: all pointer fields; null keys are skipped like missing ones.
        const data: Record<string, unknown> = {};
        if (typeof in1.title === 'string') data.title = in1.title;
        if (typeof in1.excerpt === 'string') data.excerpt = in1.excerpt;
        if (typeof in1.coverImage === 'string') data.coverImage = in1.coverImage;
        if (typeof in1.metaTitle === 'string') data.metaTitle = in1.metaTitle;
        if (typeof in1.metaDescription === 'string') data.metaDescription = in1.metaDescription;
        if (in1.isPublished === true && p.publishedAt === null) data.publishedAt = new Date();
        if (typeof in1.isPublished === 'boolean') data.isPublished = in1.isPublished;
        if (Object.keys(data).length > 0) {
          await ctx.db.blogPost.update({ where: { id: p.id }, data });
        }
        if (typeof in1.content === 'string') {
          try {
            await putObject(p.contentKey, new TextEncoder().encode(in1.content), 'text/markdown');
          } catch (e) {
            throw e instanceof StorageUnavailableError
              ? new ApiError(501, e.message)
              : new ApiError(500, e instanceof Error ? e.message : String(e));
          }
          await ctx.db.blogPost.update({
            where: { id: p.id },
            data: { contentSize: in1.content.length },
          });
        }
        const fresh = await fetchPost(ctx.db, 'id', p.id);
        return toDTO(fresh);
      }),
    )
    .handle(
      'deletePost',
      wrap(true, async (ctx) => {
        await requireAdmin(ctx);
        const p = await fetchPost(ctx.db, 'id', ctx.param('id'));
        await ctx.db.blogPost.delete({ where: { id: p.id } });
        void deleteObject(p.contentKey).catch(() => {});
        return { success: true };
      }),
    )
    .handle(
      'getPost',
      wrap(true, async (ctx) => {
        const p = await fetchPost(ctx.db, 'slug', ctx.param('slug'));
        if (ctx.query.includeContent === 'true') {
          let content: string;
          try {
            content = new TextDecoder().decode(await getObject(p.contentKey));
          } catch (e) {
            throw e instanceof StorageUnavailableError
              ? new ApiError(501, 'includeContent requires R2_* env vars')
              : new ApiError(500, e instanceof Error ? e.message : String(e));
          }
          return toDTO(p, content);
        }
        if (p.isPublished) {
          await ctx.db
            .$queryRaw`UPDATE "blog_posts" SET "viewCount" = "viewCount" + 1 WHERE "id" = ${p.id}`;
        }
        return toDTO(p);
      }),
    ),
);
